### ISSUE TITLE
Static checker does not verify hash types in many cases

### ISSUE THREAD

**Issue by XrXr** [sorbet/sorbet#11]
```
> git rev-parse HEAD
f96dc929d9cd4e10048ec4e907e659c50b7a33b7
> cat repro1.rb
# typed: true
a = T.let({symbol_key: 'integer_value'.size}, T::Hash[Integer, Symbol])
> bazel-bin/main/sorbet repro1.rb
No errors! Great job.
```
It is also not checked for method parameters:

```rb
# typed: true
extend T::Sig
  
sig {params(x: T::Hash[Integer, String]).returns(String)}
def foo(x)
  x.to_s
end

foo(symbol_key: 'integer value'.size) # An error here would be nice
```

Playground: https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%20%20%0Asig%20%7Bparams(x%3A%20T%3A%3AHash%5BInteger%2C%20String%5D).returns(String)%7D%0Adef%20foo(x)%0A%20%20x.to_s%0Aend%0A%0Afoo(symbol_key%3A%20'integer%20value'.size)%20%23%20An%20error%20here%20would%20be%20nice%0A

**Comment by XrXr** [comment]
Related: #713

**Comment by rendaw** [comment]
The issue mentioned in #200 is especially important because AFAICT there's no way to disable runtime checks (#2331).  A mis-typed hash will not cause any alerts in the static checking but will break the software due to runtime checks.

So more than great to come back to, I'd argue that either #200 or this should be high priority.

**Comment by DarkDimius** [comment]
> A mis-typed hash will not cause any alerts in the static checking but will break the software due to runtime checks

This is true about any wrong type that is called from an untyped location.
It's your call what to do in this case(see comment in the other issue you referred to where Allan describes how to tweak it). The default behaviour is what Stripe does: we prohibit unexpected behaviour. We choose to fail requests if types are wrong because the wrong types lead to behaviour that developers didn't expect to be possible.

**Comment by jaredbeck** [comment]
Copying the example from #3405 (closed) here, in case it is helpful.

#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0AT.let(%0A%20%20%7B%201%20%3D%3E%20%5B%3Ax%5D%20%7D%2C%0A%20%20T%3A%3AHash%5BInteger%2C%20T%3A%3ASet%5BSymbol%5D%5D%0A)%0A)

```ruby
# typed: true
T.let(
  { 1 => [:x] },
  T::Hash[Integer, T::Set[Symbol]]
)
```

#### Expected behavior

Sorbet should say e.g. "Expected Set, got Array"

#### Observed output

```
No errors! Great job.
```

**Comment by karmakaze** [comment]
May be partially fixed by #4497 Use inferred type for literal Hash key (always) and value (when single-classtype or nilable-single-classtype)
